node/src/components/storage.rs

```diff
@@ -1487,6 +1689,48 @@ impl Storage {
             .map_err(Into::into)
             .and_then(|mut tx| tx.get_value(self.deploy_db, &deploy_hash))
     }
+
+    /// Creates a serialized representation of a `FetchedOrNotFound` and the resulting message.
+    ///
+    /// If the given item is `Some`, returns a serialization of `FetchedOrNotFound::Fetched`. If
+    /// enabled, the given serialization is also added to the in-memory pool.
+    ///
+    /// If the given item is `None`, returns a non-pooled serialization of
+    /// `FetchedOrNotFound::NotFound`.
+    fn update_pool_and_send<REv, T>(
+        &mut self,
+        effect_builder: EffectBuilder<REv>,
+        sender: NodeId,
+        serialized_id: &[u8],
+        id: T::Id,
+        opt_item: Option<T>,
+    ) -> Result<Effects<Event>, FatalStorageError>
+    where
+        REv: From<NetworkRequest<NodeId, Message>> + Send,
+        T: Item,
+    {
+        let fetched_or_not_found = FetchedOrNotFound::from_opt(id, opt_item);
+        let serialized = fetched_or_not_found
+            .to_serialized()
+            .map_err(FatalStorageError::StoredItemSerializationFailure)?;
+        let shared: Arc<[u8]> = serialized.into();
+
+        if self.enable_mem_deduplication && fetched_or_not_found.was_found() {
+            self.serialized_item_pool
+                .put(serialized_id.into(), Arc::downgrade(&shared));
```

Comment:
I still wonder whether the key should include the tag.
I don't think it's possible right now, but if there were two items with the same hash but different tags, that could return the wrong thing, could it?

Reply:
Yes, but that'd be a hash collision, I don't think we are coding to defend against these.

The other case is that roughly the same item is valid as both, e.g. in the case of both being 0 or empty or something like that. For this to be an issue, they would have to hash to the same hash, but serialize to different representations. If that's the case, we may have a bigger issue on our hands?

Reply:
Actually, not all items are content-addressed. E.g. there's a block-with-signatures-by-height item type and a block-header-with-signatures-by-height.
Won't we store the serialized block _header_ under the same key as the serialized full block?

Reply:
Hmm, you are right.

Reply:
Yes, sadly IDs are domain specific. I've added the tag to disambiguate.